src/codebase.md

```diff
@@ -243,4 +243,40 @@ crates optimally for your semantic needs.
 | [regex](https://docs.rs/regex/)         | err, regular expressions           |
 | [ring](https://docs.rs/ring/)           | the leading crypto library         |
 | [nalgebra](https://docs.rs/nalgebra/)   | linear algebra                     |
-| [once_cell](https://docs.rs/once_cell/) | complex static data                |
\ No newline at end of file
+| [once_cell](https://docs.rs/once_cell/) | complex static data                |
+
+## How should I call C++ functions from Rust and vice-versa?
+
+Behind the scenes, Rust natively supports calling C functions using `extern "C"`,
+`#[repr(C)]` and `#[no_mangle]`. This is, of course, error prone and involves
+lots of `unsafe` code. So don't do that: there are better ways.
+
+Your initial decision should be a matter of policy. Choose between:
+
+* I want to manually specify the cross-language interface in order to keep
+  a degree of isolation and to avoid proliferation of uncontrolled
+  cross-language boundaries. Or:
+* I need to be able to use a wide variety of _existing_ C++ functions and types
+  from Rust or vice-versa, without the need to specify the interface.
+
+If you choose the former approach, the "market leader" for C++ interoperability
+is [cxx](https://cxx.rs). You specify cross-language interfaces in a
+Rust-like interface definition language (within your Rust file). Both Rust and
+C++ code is generated, to make the interface natural in each language. For
+example, you'll find idiomatic Rust wrappers for [`std::string`](https://docs.rs/cxx/1.0.50/cxx/struct.CxxString.html) and [`std::unique_ptr`](https://docs.rs/cxx/1.0.50/cxx/struct.UniquePtr.html).
+This doesn't just make things easier and more idiomatic: by understanding
+standard C++ ownership models, `cxx` can allow Rust code to interact with
+C++ without widespread use of `unsafe`.
+
+If you need to access a wide variety of existing C++ types and functions, you
+won't want to specify the interface an an IDL. Instead, you can generate bindings
```

Comment:
"You won't want to specify the interface an an IDL" is not clear to me. This sort of makes it sound like you save code/work by not describing an interface, but fwiw that is definitely not my experience. The typical binding we've converted from bindgen to cxx in my codebase has been -500 lines +300 lines.

In my experience it's not a matter of "specifying an interface" vs "not specifying an interface" -- it's a matter of "specifying a machine-comprehensible interface" vs "specifying an adhoc non-machine-comprehensible interface". In the bindgen case even more work goes into wrapping idiomatic C++ signatures into something bindgen compatible: unique ptrs to raw ptrs, translating string types to ptr/len or folly::StringPiece and back to a string type in the other language, etc.

I may be able to share these diffs with you if you are interested.

Reply:
Thanks! Do you mind if I quote you?

Reply:
Sure -- go ahead.